mlir/include/Catalyst/IR/CatalystOps.td

```diff
@@ -77,7 +78,8 @@ def PrintOp : Catalyst_Op<"print"> {
     );
 }
 
-def CustomCallOp: Catalyst_Op<"custom_call"> {
+def CustomCallOp: Catalyst_Op<"custom_call",
+    [DeclareOpInterfaceMethods<MemoryEffectsOpInterface>]> {
```

Comment:
Hmm, I know JAX has different types of callbacks, some are side-effect full and some side-effect free (purposefully separated):
https://jax.readthedocs.io/en/latest/notebooks/external_callbacks.html#flavors-of-callback

 I think the functions we invoke from lapack should be side-effect free right?

Reply:
Thinking further about this, once this op is bufferized it _would_ be adequate to model it with memory read/writes though 🤔

In the existing code, without any explicit interfaces or traits, the op is treated as conservative as possible anyway, so I don't think adding this interface as implemented changes anything. Just some food for thought I suppose.

Reply:
I'm not sure what is the outcome of this. Do you want me to delete it or not? I prefer to keep since I would like to model these functions as black boxes.

Reply:
I'm not sure either 😅 As I was thinking through it, several considerations came to mind. For example, we could have pure callbacks, as JAX has them, but that only makes sense pre-bufferization. Either way the setup is fine as is though 👍 

> I prefer to keep since I would like to model these functions as black boxes.

Just mentioning that this is already the case if you put nothing, an operation without any Trait/Interface is a black box. But I don't mind being explicit either, as long as each effect is a conscious choice (it didn't sound like this was the case, more like we were uncertain. If we are uncertain I think putting nothing reflects this better while still modelling a black box.)

Reply:
Mmm... now you made me more uncertain. :sweat_smile: I think I will delete it then. Thanks!

Reply:
From the [documentation](https://mlir.llvm.org/docs/Rationale/SideEffectsAndSpeculation/)

> When adding a new op, ask:
>
>    Does it read from or write to the heap or stack? It should probably implement MemoryEffectsOpInterface.

The line above is what makes me want to keep it. Ok, I'll just keep it.